boring/src/ssl/callbacks.rs

```diff
@@ -95,73 +103,90 @@ where
         + Sync
         + Send,
 {
-    unsafe {
-        let ssl = SslRef::from_ptr_mut(ssl);
-        let callback_idx = SslContext::cached_ex_index::<F>();
-
-        let callback = ssl
-            .ssl_context()
-            .ex_data(callback_idx)
-            .expect("BUG: psk callback missing") as *const F;
-        let identity = if identity.is_null() {
-            None
-        } else {
-            Some(CStr::from_ptr(identity).to_bytes())
-        };
-        // Give the callback mutable slices into which it can write the psk.
-        let psk_sl = slice::from_raw_parts_mut(psk as *mut u8, max_psk_len as usize);
-        match (*callback)(ssl, identity, psk_sl) {
-            Ok(psk_len) => psk_len as u32,
-            Err(e) => {
-                e.put();
-                0
-            }
+    // SAFETY: boring provides valid inputs.
+
+    let ssl = unsafe { SslRef::from_ptr_mut(ssl) };
+
+    let identity = if !identity.is_null() {
+        Some(unsafe { CStr::from_ptr(identity) }.to_bytes())
+    } else {
+        None
+    };
+
+    // Give the callback mutable slices into which it can write the psk.
+    let psk_sl = unsafe { slice::from_raw_parts_mut(psk as *mut u8, max_psk_len as usize) };
+
+    let ssl_context = ssl.ssl_context().to_owned();
+    let callback = ssl_context
+        .ex_data(SslContext::cached_ex_index::<F>())
+        .expect("BUG: psk callback missing");
+
+    match callback(ssl, identity, psk_sl) {
+        Ok(psk_len) => psk_len as u32,
+        Err(e) => {
+            e.put();
+            0
         }
     }
 }
 
-pub extern "C" fn ssl_raw_verify<F>(
+pub(super) unsafe extern "C" fn ssl_raw_verify<F>(
     preverify_ok: c_int,
     x509_ctx: *mut ffi::X509_STORE_CTX,
 ) -> c_int
 where
     F: Fn(bool, &mut X509StoreContextRef) -> bool + 'static + Sync + Send,
 {
-    unsafe {
-        let ctx = X509StoreContextRef::from_ptr_mut(x509_ctx);
-        let ssl_idx = X509StoreContext::ssl_idx().expect("BUG: store context ssl index missing");
-        let callback_idx = Ssl::cached_ex_index::<Arc<F>>();
-
-        let callback = ctx
-            .ex_data(ssl_idx)
-            .expect("BUG: store context missing ssl")
-            .ex_data(callback_idx)
-            .expect("BUG: ssl verify callback missing")
-            .clone();
-
-        callback(preverify_ok != 0, ctx) as c_int
-    }
+    // SAFETY: boring provides valid inputs.
+    let ctx = unsafe { X509StoreContextRef::from_ptr_mut(x509_ctx) };
+
+    let ssl_idx = X509StoreContext::ssl_idx().expect("BUG: store context ssl index missing");
+
+    // NOTE(nox): I'm pretty sure this Arc<F> is unnecessary here as there is
+    // no way to get a `&mut SslRef` from a `&mut X509StoreContextRef`, and I
+    // don't understand how this callback is different from `raw_verify` above.
+    let callback = ctx
```

Comment:
shouldn't we obtain strong ref here?

Reply:
No, because the callback is owned as it is an `Arc<F>`. There is no unsafe code here to get `callback` so it wouldn't type check if it was incorrect anyway.